File: test-cases/full-stack/hard/junction/v1.0.0/reference-impl/base/src/menus.ts
// Junction — menu definitions (specs/flow.md "Required menus", DESIGN §4, §5.2).
//
// A single source of truth for each state's menu items so the renderer draws them and the
// input layer's keyboard navigation (↑/↓ + Enter) drive the same list. Mirrors valence's
// `menus.ts`: the render and input slices agree on the item order and actions here.

import type { GameState } from "./types";
import type { Game } from "./sim";

export interface MenuItem {
  label: string;
  action: string;
}

// The navigable menu for a state; empty for the in-play states that have no list menu.
export function menuItems(state: GameState, game: Game): MenuItem[] {
  switch (state) {
    case "title":
      return [
        { label: game.mode.menuLabel, action: "menu:play" }, // "NEW CITY"
        { label: "HOW TO PLAY", action: "menu:howto" },
      ];
    case "howto":
      return [{ label: "BACK", action: "menu:back" }];
    case "paused":
      return [
        { label: "RESUME", action: "menu:resume" },
        { label: "RESTART", action: "menu:restart" },
        { label: "QUIT TO MENU", action: "menu:quit" },
      ];
    case "bankrupt":
      return [
        { label: "TRY AGAIN", action: "menu:again" },
        { label: "MENU", action: "menu:menu" },
      ];
    default:
      return [];
  }
}

File: test-cases/full-stack/medium/junction/v1.0.0/reference-impl/base/sim-core/src/menus.rs
// Junction — menu definitions (specs/flow.md "Required menus", DESIGN §4, §5.2), ported
// from `menus.ts`.
//
// The single source of truth for each state's menu items. Under the Rust/wasm boundary the
// simulation core OWNS the menus and the keyboard-nav index (specs/simulation.md): the front
// end reads the list to draw it, moves the highlight with the arrow keys / pointer, and
// confirms an item — all through the boundary — while the item's action is dispatched here.

use crate::mode::MODE;
use crate::types::GameState;

pub struct MenuItem {
    pub label: String,
    pub action: &'static str,
}

/// The navigable menu for a state; empty for the in-play states that have no list menu.
pub fn menu_items(state: GameState) -> Vec<MenuItem> {
    match state {
        GameState::Title => vec![
            MenuItem { label: MODE.menu_label.to_string(), action: "menu:play" },
            MenuItem { label: "HOW TO PLAY".to_string(), action: "menu:howto" },
        ],
        GameState::Howto => vec![MenuItem { label: "BACK".to_string(), action: "menu:back" }],
        GameState::Paused => vec![
            MenuItem { label: "RESUME".to_string(), action: "menu:resume" },
            MenuItem { label: "RESTART".to_string(), action: "menu:restart" },
            MenuItem { label: "QUIT TO MENU".to_string(), action: "menu:quit" },
        ],
        GameState::Bankrupt => vec![
            MenuItem { label: "TRY AGAIN".to_string(), action: "menu:again" },
            MenuItem { label: "MENU".to_string(), action: "menu:menu" },
        ],
        GameState::Playing => Vec::new(),
    }
}
